### ISSUE TITLE
Accessibility textScaleFactor doesn't affect fixed-sized widgets.

### ISSUE THREAD

**Issue by gspencergoog** [flutter/flutter#12311]
When the `MediaQuery` `textScaleFactor` is set to something larger than 1.0, there can be rendering problems (clipped text) with widgets that have fixed sizes.  Basically anything that the [Material Design spec](https://material.io/guidelines/layout/metrics-keylines.html) calls out as having a fixed size in `dp` (device-independent pixels), but contains text that is specified in `sp` (scaled pixels) has this problem.

For instance, the spec calls out toolbars (a la AppBar) as being a fixed `56dp` in height, but if the system's text scale is set to (say) 1.5, then most titles would be too large to fit in their assigned space, and the toolbar title will be cropped and unreadable.

For these cases, I have asked the Material Design team if should we just ignore the `textScaleFactor` and render with the size in the spec for widgets with a specified size, or if the sizes of such elements should be considered to be in "sp" instead of "dp", and scale them when the text scales?

Note that if we ignore the scale in fixed-sized widgets, then there is a good likelihood that no text at all will scale in an application's UI (if it is made up entirely of exempted widgets). However, if we scale only the widgets that contain text, it could wreak havoc on the layout in general.

The quote "To enable system font size in an Android app, mark text and their associated containers to be measured in scaleable pixels (sp) ." [in the spec](https://material.io/guidelines/usability/accessibility.html#accessibility-style) seems to imply that containers should scale with their contents, but I'm not sure if that is meant to only apply to containers without specified fixed sizes or not.

In any case, we need to do something, and neither choice is a good one. I would prefer to resize all text-containing widgets by the amount needed to accommodate the larger (or smaller) text, but that is a significant change, and will probably introduce significant layout problems.

One idea I had was to somehow do layout in a context that is scaled by `1/textScaleFactor`, and then scale it by `textScaleFactor` again before rendering.  This would probably result in the least disruption to layout, but would scale everything in the context, not just text.  It would be like laying out for a device with lower resolution, but rendering it at full resolution.

**Comment by gspencergoog** [comment]
I also just realized that in almost all cases, we only need to scale in height, which simplifies things somewhat, although leads to some crappy layout. (I think some crappy layout is inevitable here.)

The ideal (in visual appearance, not performance) scaling algorithm would traverse all the children of a widget, finding any text that honors greater-than-one scales, grow them, see if they exceed the heights of any (text and non-text) siblings, if any, and increase the height of their container by that many pixels. Then it would move up a level, and see if that container exceeded the height of any other siblings, and so on until reaching the root.  This way we wouldn't grow more than needed.

Scales smaller than one are generally not as much of a problem, although if the text was the largest item in the container, then the container should probably shrink by the number of pixels needed to wrap the next largest item in the container or the shrunken text, whichever is largest.

**Comment by gspencergoog** [comment]
OK, the Material Design folks say that ideally we should allow text to expand to the point where it would be clipped vertically, without changing the dimensions of the fixed-sized container. As soon as it would be clipped, expand the fixed-sized container just enough to contain the expanded text.

This is basically what I was alluding to above.  I don't feel confident that a general purpose algorithm will fit all cases, so this feels like a yak shave is in order to address each of the fixed-sized containers to have this behavior.

**Comment by Hixie** [comment]
Clearing assignee since Greg has moved on to other things.

**Comment by gspencergoog** [comment]
Geez, sounds like I left or something. :-)

**Comment by Hixie** [comment]
All I know is I can't see you at your desk. :-P

**Comment by jonahwilliams** [comment]
Since we've decided to deal with this individually, lets file and resolve individual issues
